src/Microsoft.IdentityModel.Tokens/TokenUtilities.cs

```diff
@@ -55,5 +55,72 @@ internal static IEnumerable<SecurityKey> GetAllSigningKeys(TokenValidationParame
                 foreach (SecurityKey key in validationParameters.IssuerSigningKeys)
                     yield return key;
         }
+
+        /// <summary>
+        /// Merges claims. If a claim with same type exists in both <paramref name="claims"/> and <paramref name="subjectClaims"/>, the one in claims will be kept.
+        /// </summary>
+        /// <param name="claims"> Collection of <see cref="Claim"/>'s.</param>
+        /// <param name="subjectClaims"> Collection of <see cref="Claim"/>'s.</param>
+        /// <returns> A Merged list of <see cref="Claim"/>'s.</returns>
+        internal static IEnumerable<Claim> MergeClaims(IEnumerable<Claim> claims, IEnumerable<Claim> subjectClaims)
+        {
+            if (claims == null)
+                return subjectClaims;
+
+            if (subjectClaims == null)
+                return claims;
+
+            List<Claim> result = claims.ToList();
+
+            foreach (Claim claim in subjectClaims)
+            {
+                if (!claims.Where(i => i.Type == claim.Type).Any())
+                    result.Add(claim);
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Gets the value type of the <see cref="Claim"/> from its value <paramref name="value"/>
+        /// </summary>
+        /// <param name="value"> The <see cref="Claim"/> value.</param>
+        /// <returns> The value type of the <see cref="Claim"/>.</returns>
+        internal static string GetClaimValueTypeFromValue(object value)
+        {
+            if (value != null)
+            {
+                if (value is string)
+                    return ClaimValueTypes.String;
+
+                if (value.GetType().Name == typeof(bool).Name)
+                    return ClaimValueTypes.Boolean;
+
+                if (value.GetType().Name == typeof(int).Name)
+                    return ClaimValueTypes.Integer32;
+
+                if (value.GetType().Name == typeof(long).Name)
+                    return ClaimValueTypes.Integer64;
+
+                if (value.GetType().Name == typeof(double).Name)
+                    return ClaimValueTypes.Double;
+
+                if (value.GetType().Name == typeof(DateTime).Name)
+                    return ClaimValueTypes.DateTime;
+
+                if (value.GetType().Name == typeof(List<>).Name)
+                {
+                    foreach (var item in (IList)value)
+                    {
+                        // What to do if the value type of objects in the list are different.
+                        return GetClaimValueTypeFromValue(item);
```

Comment:
given you are not actually looking to figure out if there are different types perhaps write the code to simply look only at the first element. not sure if it's worth the effort to go through all elements here, depends on what the use case is.

Reply:
perhaps you should return typeof(List<>).Name) as your call pattern to this has to do special handling here?

---
In reply to: [428331612](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1417#discussion_r428331612) [](ancestors = 428331612)